### PR TITLE
Add unit to arguments

### LINKED ISSUE #673: unit parameter
### Versions

GitHub master

### Description

When using pymodbus within an IDE like Spyder3, some ModbusSerialClient functions contain a 'hidden' parameter, namely one that does not appear as autocompletion, and that can only be found by looking in the docstrings by opening the source code. I am referring to parameter 'unit' in https://github.com/riptideio/pymodbus/blob/master/pymodbus/repl/client/mclient.py
image

It's only in the docstrings that one can find about how this parameter can be tuned. I needed to tune it from default's 0 to 1 and could not find how in the beginning.

image

I suggest to change the function declaration as follows:

`def read_holding_registers(self, address, count=1, unit=0, **kwargs):`

There are several more functions with that hidden parameter in that file.

### PR DESCRIPTION
**Pull request by pkubanek**
Fix #673. Adds unit to argument list for selected functions.

### CODE REVIEW

File: pymodbus/repl/client/mclient.py [hunk 1]
```diff
@@ -109,107 +109,114 @@ def read_coils(self, address, count=1, **kwargs):
...
         return ExtendedRequestSupport._process_exception(resp)
 
     @handle_brodcast
-    def write_coil(self, address, value, **kwargs):
+    def write_coil(self, address, value, unit=0, **kwargs):
         """Write `value` to coil at `address`.
 
         :param address: coil offset to write to
         :param value: bit value to write
+        :param unit: Modbus slave unit ID
```
Comment (janiversen) [inline pymodbus/repl/client/mclient.py:134]: Please use identical description for unit in all places, using different descriptions are confusion.

I prefer this description overall.

**Review by janiversen — CHANGES_REQUESTED** [review]
Seems you changed mind, about halfway through the modifications and started to use another description, please correct the first half to the latter.

**Comment by janiversen** [comment]
Der er et pylint problem
```
pymodbus/repl/client/mclient.py:99:43: W0613: Unused argument 'unit' (unused-argument)
[595](https://github.com/riptideio/pymodbus/runs/7876551304?check_suite_focus=true#step:5:596)
pymodbus/repl/client/mclient.py:300:18: E1121: Too many positional arguments for constructor call (too-many-function-args)
[596](https://github.com/riptideio/pymodbus/runs/7876551304?check_suite_focus=true#step:5:597)
************* Module pymodbus.client.mixin
[597](https://github.com/riptideio/pymodbus/runs/7876551304?check_suite_focus=true#step:5:598)
pymodbus/client/mixin.py:160:55: W0613: Unused argument 'unit' (unused-argument)
```

File: pymodbus/repl/client/mclient.py (hunk 1 above)
Comment (pkubanek) [reply pymodbus/repl/client/mclient.py:136]: Can you be please more specific? Apart from unit iD vs. unit ID, I did not see the difference. Thanks

Comment (janiversen) [reply pymodbus/repl/client/mclient.py:136]: ???? You did that in an earlier commit.

File: pymodbus/register_read_message.py [hunk 2]
```diff
@@ -10,13 +11,14 @@ class ReadRegistersRequestBase(ModbusRequest):
...
-    def __init__(self, address, count, **kwargs):
+    def __init__(self, address, count, unit=Defaults.UnitId, **kwargs):
         """Initialize a new instance.
 
         :param address: The address to start the read from
         :param count: The number of registers to read
+        :param unit: Modbus slave unit ID
         """
-        ModbusRequest.__init__(self, **kwargs)
+        ModbusRequest.__init__(self, unit, **kwargs)
```
Comment (janiversen) [inline pymodbus/register_read_message.py:21]: Please change to:
```suggestion
        super().__init__(self, unit, **kwargs)
```

File: pymodbus/register_read_message.py [hunk 3]
```diff
@@ -57,12 +59,13 @@ class ReadRegistersResponseBase(ModbusResponse):
...
 
-    def __init__(self, values, **kwargs):
+    def __init__(self, values, unit=Defaults.UnitId, **kwargs):
         """Initialize a new instance.
 
         :param values: The values to write to
+        :param unit: Modbus slave unit ID
         """
-        ModbusResponse.__init__(self, **kwargs)
+        ModbusResponse.__init__(self, unit, **kwargs)
```
Comment (janiversen) [inline pymodbus/register_read_message.py:68]:
```suggestion
        super().__init__(self, unit, **kwargs)
```

File: pymodbus/register_read_message.py [hunk 4]
```diff
@@ -115,13 +118,14 @@ class ReadHoldingRegistersRequest(ReadRegistersRequestBase):
...
-    def __init__(self, address=None, count=None, **kwargs):
+    def __init__(self, address=None, count=None, unit=Defaults.UnitId, **kwargs):
         """Initialize a new instance of the request.
 
         :param address: The starting address to read from
         :param count: The number of registers to read from address
+        :param unit: Modbus slave unit ID
         """
-        ReadRegistersRequestBase.__init__(self, address, count, **kwargs)
+        ReadRegistersRequestBase.__init__(self, address, count, unit, **kwargs)
```
Comment (janiversen) [inline pymodbus/register_read_message.py:128]:
```suggestion
        super().__init__(self, address, count, unit, **kwargs)
```

File: pymodbus/register_read_message.py [hunk 5]
```diff
@@ -171,13 +175,14 @@ class ReadInputRegistersRequest(ReadRegistersRequestBase):
...
```
Comment (janiversen) [inline pymodbus/register_read_message.py:185]:
```suggestion
        super().__init__(self, address, count, unit, **kwargs)
```

**Review by janiversen — CHANGES_REQUESTED** [review]
update.

**Comment by janiversen** [comment]
I wonder what you broke, but I will take a closer look.

File: pymodbus/pdu.py [hunk 6]
```diff
@@ -94,9 +98,12 @@ def calculateRtuFrameSize(cls, buffer):  # pylint: disable=invalid-name
...
 
-    def __init__(self, **kwargs):
-        """Proxy to the lower level initializer."""
-        ModbusPDU.__init__(self, **kwargs)
+    def __init__(self, unit=Defaults.UnitId, **kwargs):
+        """Proxy to the lower level initializer.
+
+        :param unit: Modbus slave unit ID
+        """
+        ModbusPDU.__init__(self, unit, **kwargs)
```
[the remark this answers is not in the document]
Comment (pkubanek) [reply pymodbus/pdu.py:106]: Yes, I looked onto that..will change it, thanks for confirming I shall change it (it is that way in upstream, time to change to Python 3.x super()..)

[the remark this answers is not in the document]
Comment (pkubanek) [reply pymodbus/pdu.py:106]: Well you need to delete self from argument list as well..

**Comment by pkubanek** [comment]
Added commit to use super() in files I modified.

File: pymodbus/pdu.py [hunk 7]
```diff
@@ -127,9 +134,14 @@ class ModbusResponse(ModbusPDU):
...
-        """Proxy the lower level initializer."""
-        ModbusPDU.__init__(self, **kwargs)
+    def __init__(self, unit=Defaults.UnitId, **kwargs):
+        """Proxy the lower level initializer.
+
+        :param unit: Modbus slave unit ID
+
+        """
+        super().__init__(unit, **kwargs)
+
```
Comment (janiversen) [inline pymodbus/pdu.py:144]:
```suggestion
```

**Review by janiversen — CHANGES_REQUESTED** [review]
flake8

**Comment by swamper123** [comment]
Ci says `pymodbus/pdu.py:146:5: E303 too many blank lines (2)`

**Comment by janiversen** [comment]
yes already solved.

**Comment by janiversen** [comment]
Thanks for the PR, hope you will want to continue.

**Pull request merged**